### PR TITLE
Misc improvements

### PR DESCRIPTION
**Pull request by nivemaham**

### CODE REVIEW

File: src/main/java/org/radarbase/management/web/rest/ProjectResource.java [hunk 1]
```diff
@@ -261,7 +262,12 @@ public ResponseEntity getAllSourcesForProject(@ApiParam Pageable pageable,
                     Boolean minimized) throws NotAuthorizedException {
         log.debug("REST request to get all Sources");
         ProjectDTO projectDto = projectService.findOneByName(projectName);
-        checkPermissionOnProject(getJWT(servletRequest), SOURCE_READ, projectDto.getProjectName());
+        RadarToken jwt = getJWT(servletRequest);
+        checkPermissionOnProject(jwt, SOURCE_READ, projectDto.getProjectName());
+        if (jwt.hasAuthority(PARTICIPANT)) {
+            throw new NotAuthorizedException(String.format("Client %s does not have "
```
Comment (nivemaham) [inline src/main/java/org/radarbase/management/web/rest/ProjectResource.java:268]: Would it be better to rephrase the mesage that a PARTICIPANT cannot perform this operation. We are not checking for permission here?

File: src/main/java/org/radarbase/management/web/rest/RoleResource.java [hunk 2]
```diff
@@ -135,6 +138,7 @@
     @GetMapping("/roles/{projectName:" + Constants.ENTITY_ID_REGEX + "}/{authorityName:"
             + Constants.ENTITY_ID_REGEX + "}")
     @Timed
+    @Secured({AuthoritiesConstants.PROJECT_ADMIN, AuthoritiesConstants.SYS_ADMIN})
```
Comment (nivemaham) [inline src/main/java/org/radarbase/management/web/rest/RoleResource.java:141]: I think except this call, all the others should be limited to SYS_ADMIN only.

File: src/main/java/org/radarbase/management/web/rest/SourceDataResource.java [hunk 3]
```diff
@@ -71,7 +73,7 @@
     public ResponseEntity<SourceDataDTO> createSourceData(@Valid @RequestBody SourceDataDTO
             sourceDataDto) throws URISyntaxException, NotAuthorizedException {
         log.debug("REST request to save SourceData : {}", sourceDataDto);
-        checkPermission(getJWT(servletRequest), SOURCEDATA_CREATE);
+        checkAuthorityAndPermission(getJWT(servletRequest), SYS_ADMIN, SOURCEDATA_CREATE);
```
Comment (nivemaham) [inline src/main/java/org/radarbase/management/web/rest/SourceDataResource.java:76]: Just to confirm, would this disregard if the action is performed by a service with client_credentials?

File: src/main/java/org/radarbase/management/web/rest/SourceResource.java [hunk 4]
```diff
@@ -77,7 +85,9 @@
     public ResponseEntity<SourceDTO> createSource(@Valid @RequestBody SourceDTO sourceDto)
             throws URISyntaxException, NotAuthorizedException {
         log.debug("REST request to save Source : {}", sourceDto);
-        checkPermission(getJWT(servletRequest), SOURCE_CREATE);
+        MinimalProjectDetailsDTO project = sourceDto.getProject();
+        String projectName = project != null ? project.getProjectName() : null;
+        checkPermissionOnProject(getJWT(servletRequest), SOURCE_CREATE, projectName);
```
Comment (nivemaham) [inline src/main/java/org/radarbase/management/web/rest/SourceResource.java:90]: This means a source cannot be created without being assigned to a project? Functionally sounds good to me. We may have to make the project required on the UI under entities/sources/create new source (until we remove that feature).

File: src/main/java/org/radarbase/management/service/SourceService.java [hunk 5]
```diff
@@ -208,24 +212,41 @@ public MinimalSourceDetailsDTO safeUpdateOfAttributes(Source sourceToUpdate,
...
+        Source existingSource = existingSourceOpt.get();
+        String project = existingSource.getProject() != null
+                ? existingSource.getProject().getProjectName()
+                : null;
+        String user = (existingSource.getSubject() != null
+                && existingSource.getSubject().getUser() != null)
+                ? existingSource.getSubject().getUser().getLogin()
+                : null;
+
+        checkPermissionOnSource(jwt, SOURCE_UPDATE, project, user, existingSource.getSourceName());
```
Comment (nivemaham) [inline src/main/java/org/radarbase/management/service/SourceService.java:234]: Did we already have these methods (checkPermissionOnSource and checkPermissionOnSubject) available in the radar-auth?

**Review by nivemaham — COMMENTED** [review]
I can't approve the PR myself. But mostly looks good to me. I left some minor comments and clarification questions.

File: radar-auth/src/main/java/org/radarbase/auth/token/AbstractRadarToken.java [hunk 6]
```diff
@@ -84,10 +85,14 @@ protected boolean hasAuthorityForPermission(Permission permission) {
      * @return {@code true} if any authority contains the permission, {@code false} otherwise
      */
     protected boolean hasAuthorityForProject(Permission permission, String projectName) {
-        return getRoles()
-                .getOrDefault(projectName, emptyList()).stream()
-                .anyMatch(permission::isAuthorityAllowed)
-                || hasNonProjectRelatedAuthorityForPermission(permission);
+        if (hasNonProjectRelatedAuthorityForPermission(permission)) {
```
[the remark this answers is not in the document]
Comment (blootsvoets) [reply radar-auth/src/main/java/org/radarbase/auth/token/AbstractRadarToken.java:87]: Indeed, it's only called internally by `hasPermission.*` which checks for client credentials already.

File: src/main/java/org/radarbase/management/service/SourceService.java (hunk 5 above)
Comment (blootsvoets) [reply src/main/java/org/radarbase/management/service/SourceService.java:234]: These have been available quite some time, but they were introduced later than the other permission checks.

File: src/main/java/org/radarbase/management/web/rest/SourceDataResource.java (hunk 3 above)
Comment (blootsvoets) [reply src/main/java/org/radarbase/management/web/rest/SourceDataResource.java:76]: Good find. It didn't, but it does now.

File: src/main/java/org/radarbase/management/web/rest/SourceResource.java (hunk 4 above)
Comment (blootsvoets) [reply src/main/java/org/radarbase/management/web/rest/SourceResource.java:90]: If the project is not valid (like `null`), `checkPermissionOnProject` will only allow sys admins and oauth clients. That is fine for this use case.

**Review by blootsvoets — APPROVED** [review]
Works -- thanks!

**Pull request merged**